File: demo_self_growing_xor/cpp/baseline_xor.cpp
// Self-growing XOR baseline in C++ with Eigen
// Equivalent to NOMA demo: no biases, Adam optimizer, same architecture
// Compile: g++ -O3 -I/path/to/eigen -o baseline_xor baseline_xor.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <random>
#include <chrono>

// Simple matrix class (no Eigen dependency for portability)
struct Matrix {
    std::vector<double> data;
    int rows, cols;
    
    Matrix(int r, int c) : rows(r), cols(c), data(r * c, 0.0) {}
    
    double& operator()(int i, int j) { return data[i * cols + j]; }
    double operator()(int i, int j) const { return data[i * cols + j]; }
    
    Matrix T() const {
        Matrix result(cols, rows);
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                result(j, i) = (*this)(i, j);
        return result;
    }
};

// Matrix operations
Matrix matmul(const Matrix& A, const Matrix& B) {
    Matrix C(A.rows, B.cols);
    for (int i = 0; i < A.rows; i++)
        for (int j = 0; j < B.cols; j++)
            for (int k = 0; k < A.cols; k++)
                C(i, j) += A(i, k) * B(k, j);
    return C;
}

Matrix sigmoid(const Matrix& A) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = 1.0 / (1.0 + std::exp(-A.data[i]));
    return result;
}

Matrix operator-(const Matrix& A, const Matrix& B) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = A.data[i] - B.data[i];
    return result;
}

Matrix operator*(const Matrix& A, const Matrix& B) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = A.data[i] * B.data[i];
    return result;
}

Matrix operator*(const Matrix& A, double s) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = A.data[i] * s;
    return result;
}

Matrix operator/(const Matrix& A, const Matrix& B) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = A.data[i] / B.data[i];
    return result;
}

Matrix operator+(const Matrix& A, double s) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = A.data[i] + s;
    return result;
}

Matrix sqrt_m(const Matrix& A) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = std::sqrt(A.data[i]);
    return result;
}

Matrix pow_m(const Matrix& A, double p) {
    Matrix result(A.rows, A.cols);
    for (size_t i = 0; i < A.data.size(); i++)
        result.data[i] = std::pow(A.data[i], p);
    return result;
}

double mean(const Matrix& A) {
    double sum = 0.0;
    for (double v : A.data) sum += v;
    return sum / A.data.size();
}

// Adam optimizer state
struct AdamState {
    Matrix m, v;
    int t = 0;
    
    AdamState(int rows, int cols) : m(rows, cols), v(rows, cols) {}
    
    void update(Matrix& param, const Matrix& grad, double lr, 
                double beta1 = 0.9, double beta2 = 0.999, double eps = 1e-8) {
        t++;
        for (size_t i = 0; i < param.data.size(); i++) {
            m.data[i] = beta1 * m.data[i] + (1.0 - beta1) * grad.data[i];
            v.data[i] = beta2 * v.data[i] + (1.0 - beta2) * grad.data[i] * grad.data[i];
            double m_hat = m.data[i] / (1.0 - std::pow(beta1, t));
            double v_hat = v.data[i] / (1.0 - std::pow(beta2, t));
            param.data[i] -= lr * m_hat / (std::sqrt(v_hat) + eps);
        }
    }
    
    // Reset for new shape (growth)
    void reset(int rows, int cols) {
        m = Matrix(rows, cols);
        v = Matrix(rows, cols);
        t = 0;
    }
};

// Random normal initialization
void rand_normal(Matrix& M, double std, std::mt19937& rng) {
    std::normal_distribution<double> dist(0.0, std);
    for (double& v : M.data) v = dist(rng);
}

int main(int argc, char* argv[]) {
    std::string csv_path = argc > 1 ? argv[1] : "out/cpp_loss.csv";
    std::ofstream csv(csv_path);
    csv << "step,loss,hidden\n";
    
    std::mt19937 rng(42);
    
    // XOR dataset
    Matrix X(4, 2), Y(4, 1);
    X(0,0)=0; X(0,1)=0; Y(0,0)=0;
    X(1,0)=0; X(1,1)=1; Y(1,0)=1;
    X(2,0)=1; X(2,1)=0; Y(2,0)=1;
    X(3,0)=1; X(3,1)=1; Y(3,0)=0;
    
    // Phase 1: hidden=2
    int hidden = 2;
    Matrix W1(2, hidden), W2(hidden, 1);
    rand_normal(W1, 0.5, rng);
    rand_normal(W2, 0.5, rng);
    
    AdamState adam_W1(2, hidden), adam_W2(hidden, 1);
    
    int step = 0;
    double lr1 = 0.05;
    
    // Phase 1: 200 iterations
    for (int iter = 0; iter < 200; iter++) {
        // Forward
        Matrix h = sigmoid(matmul(X, W1));
        Matrix pred = sigmoid(matmul(h, W2));
        Matrix err = pred - Y;
        double loss = mean(err * err);
        
        csv << step << "," << loss << "," << hidden << "\n";
        step++;
        
        // Backward
        Matrix dloss = err * (2.0 / X.rows);
        Matrix dout = dloss * pred * (pred * (-1.0) + 1.0);
        Matrix dW2 = matmul(h.T(), dout);
        
        Matrix dh = matmul(dout, W2.T());
        Matrix dh_act = dh * h * (h * (-1.0) + 1.0);
        Matrix dW1 = matmul(X.T(), dh_act);
        
        // Adam updates
        adam_W1.update(W1, dW1, lr1);
        adam_W2.update(W2, dW2, lr1);
    }
    
    // Growth: hidden 2 -> 16
    int hidden_big = 16;
    Matrix W1_new(2, hidden_big), W2_new(hidden_big, 1);
    rand_normal(W1_new, 0.3, rng);
    rand_normal(W2_new, 0.3, rng);
    
    // Copy old weights
    for (int i = 0; i < 2; i++)
        for (int j = 0; j < hidden; j++)
            W1_new(i, j) = W1(i, j);
    for (int i = 0; i < hidden; i++)
        W2_new(i, 0) = W2(i, 0);
    
    W1 = W1_new;
    W2 = W2_new;
    hidden = hidden_big;
    
    // RESET Adam state (this is the plumbing cost!)
    adam_W1.reset(2, hidden);
    adam_W2.reset(hidden, 1);
    
    double lr2 = 0.12;
    
    // Phase 2: 342 iterations
    for (int iter = 0; iter < 342; iter++) {
        // Forward
        Matrix h = sigmoid(matmul(X, W1));
        Matrix pred = sigmoid(matmul(h, W2));
        Matrix err = pred - Y;
        double loss = mean(err * err);
        
        csv << step << "," << loss << "," << hidden << "\n";
        step++;
        
        // Backward
        Matrix dloss = err * (2.0 / X.rows);
        Matrix dout = dloss * pred * (pred * (-1.0) + 1.0);
        Matrix dW2 = matmul(h.T(), dout);
        
        Matrix dh = matmul(dout, W2.T());
        Matrix dh_act = dh * h * (h * (-1.0) + 1.0);
        Matrix dW1 = matmul(X.T(), dh_act);
        
        // Adam updates
        adam_W1.update(W1, dW1, lr2);
        adam_W2.update(W2, dW2, lr2);
    }
    
    csv.close();
    std::cout << "[cpp] wrote " << csv_path << " (total steps=" << step << ")\n";
    std::cout << "[cpp] note: growth RESETS optimizer state — unlike NOMA demo\n";
    
    return 0;
}

File: demo_self_growing_xor/python/baseline_xor.py
#!/usr/bin/env python3
"""
Minimal Python baseline for XOR.
- Phase 1: tiny hidden layer (2 units) with vanilla SGD
- Phase 2: manual growth to 16 units that *resets* optimizer state to show the plumbing cost
Outputs: csv at path provided as first arg (default: out/python_loss.csv)
"""
from __future__ import annotations

import csv
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Tuple

import numpy as np

RNG = np.random.default_rng(42)  # seed for reproducibility


def sigmoid(x: np.ndarray) -> np.ndarray:
    return 1.0 / (1.0 + np.exp(-x))


def forward(x: np.ndarray, W1: np.ndarray, b1: np.ndarray, W2: np.ndarray, b2: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
    h = sigmoid(x @ W1 + b1)
    out = sigmoid(h @ W2 + b2)
    return h, out


def mse(pred: np.ndarray, target: np.ndarray) -> float:
    diff = pred - target
    return float(np.mean(diff * diff))


@dataclass
class AdamState:
    """Adam optimizer state for a single parameter."""
    m: np.ndarray  # first moment
    v: np.ndarray  # second moment
    t: int = 0     # timestep

    @classmethod
    def zeros_like(cls, arr: np.ndarray) -> "AdamState":
        return cls(m=np.zeros_like(arr), v=np.zeros_like(arr), t=0)


def adam_update(param: np.ndarray, grad: np.ndarray, state: AdamState, 
                lr: float = 0.01, beta1: float = 0.9, beta2: float = 0.999, eps: float = 1e-8) -> Tuple[np.ndarray, AdamState]:
    state.t += 1
    state.m = beta1 * state.m + (1 - beta1) * grad
    state.v = beta2 * state.v + (1 - beta2) * (grad ** 2)
    m_hat = state.m / (1 - beta1 ** state.t)
    v_hat = state.v / (1 - beta2 ** state.t)
    param = param - lr * m_hat / (np.sqrt(v_hat) + eps)
    return param, state


def train_phase_adam(x: np.ndarray, y: np.ndarray, steps: int, lr: float, hidden: int,
                     W1: np.ndarray, W2: np.ndarray,
                     adam_states: dict | None = None) -> Tuple[Tuple[np.ndarray, np.ndarray], list[float], dict]:
    """Train with Adam optimizer (no biases, matching NOMA). Returns updated params, losses, and adam states."""
    if adam_states is None:
        adam_states = {
            "W1": AdamState.zeros_like(W1),
            "W2": AdamState.zeros_like(W2),
        }
    
    losses: list[float] = []
    for _ in range(steps):
        # Forward (no biases, like NOMA)
        h = sigmoid(x @ W1)
        out = sigmoid(h @ W2)
        
        # Compute loss
        diff = out - y
        loss = float(np.mean(diff * diff))
        losses.append(loss)
        
        # Backward
        dloss = 2.0 * diff / x.shape[0]
        dout = dloss * out * (1.0 - out)
        dW2 = h.T @ dout
        
        dh = dout @ W2.T
        dh_act = dh * h * (1.0 - h)
        dW1 = x.T @ dh_act
        
        # Adam updates
        W1, adam_states["W1"] = adam_update(W1, dW1, adam_states["W1"], lr)
        W2, adam_states["W2"] = adam_update(W2, dW2, adam_states["W2"], lr)
    
    return (W1, W2), losses, adam_states


def main() -> int:
    out_path = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("out/python_loss.csv")
    out_path.parent.mkdir(parents=True, exist_ok=True)

    X = np.array([[0.0, 0.0], [0.0, 1.0], [1.0, 0.0], [1.0, 1.0]])
    Y = np.array([[0.0], [1.0], [1.0], [0.0]])

    # Phase 1: tiny network (hidden=2), expect plateau — SAME AS NOMA
    hidden_small = 2
    W1 = RNG.normal(scale=0.5, size=(2, hidden_small))
    W2 = RNG.normal(scale=0.5, size=(hidden_small, 1))

    params, loss1, adam_states = train_phase_adam(X, Y, steps=200, lr=0.05, hidden=hidden_small,
                                                   W1=W1, W2=W2, adam_states=None)

    # Phase 2: grow to 16 hidden units — MUST RESET ADAM STATE (this is the pain point!)
    hidden_big = 16
    W1_old, W2_old = params
    
    # Create new larger arrays and copy old weights
    W1_new = RNG.normal(scale=0.3, size=(2, hidden_big))
    W1_new[:, :hidden_small] = W1_old

    W2_new = RNG.normal(scale=0.3, size=(hidden_big, 1))
    W2_new[:hidden_small, :] = W2_old

    # RESET Adam state — this is what NOMA avoids!
    params2, loss2, _ = train_phase_adam(X, Y, steps=342, lr=0.12, hidden=hidden_big,
                                          W1=W1_new, W2=W2_new,
                                          adam_states=None)  # <-- RESET!

    # Write CSV with hidden column
    with out_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow(["step", "loss", "hidden"])
        step = 0
        for loss in loss1:
            writer.writerow([step, f"{loss:.6f}", hidden_small])
            step += 1
        for loss in loss2:
            writer.writerow([step, f"{loss:.6f}", hidden_big])
            step += 1

    print(f"[baseline] wrote {out_path} (total steps={step})")
    print("[baseline] note: growth RESETS optimizer state — unlike NOMA demo")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
